towhee/dataframe/iterators.py

```diff
@@ -0,0 +1,227 @@
+            (`None`)
+                In the case that the `DataFrame` is not sealed and the new rows are
+                not ready yet, the iterator will return `None`. The caller should
+                determine whether to block the iteration or exit the loop.
+        Raises:
+            (`StopIteration`)
+                The iteration end iff the `DataFrame` is sealed and the last row is
+                reached.
+        """
+        if self._done:
+            raise StopIteration
+
+        df = self._df_ref()
+
+        code, row = df.get(self._offset, count = 1, iter_id = self._id)
+
+        if code == Responses.INDEX_GC:
+            raise IndexError
+
+        elif code == Responses.INDEX_OOB_UNSEALED:
+            if self._block:
+                cv = df.notify_block(self._id, self._offset, 1)
+                with cv:
+                    cv.wait()
+
+                return self.__next__()
+
+            return None
+
+        elif code == Responses.APPROVED_CONTINUE:
+            # subtract one due to step.
+            df.ack(self._id, self._offset)
+            self._offset += 1
+            return row
+
+        elif code == Responses.INDEX_OOB_SEALED:
+            raise StopIteration
+
+        elif code == Responses.APPROVED_DONE:
+            self._done = True
+            df.ack(self._id, self._offset)
+            self._offset += 1
+            return row
+
+        elif code == Responses.KILLED:
+            raise StopIteration
+
+        else: # 'unkown_error'
+            raise Exception
+
+    @property
+    def id(self):
+        return self._id
+
+
+class BatchIterator:
+    """
+    A row-based batch `DataFrame` iterator.
+    """
+    def __init__(self, df: DataFrame, batch_size = 1, step = 1, block = False):
+        self._df_ref = weakref.ref(df)
+        self._offset = 0
+        self._batch_size = batch_size
+        self._step = step
+        self._block = block
+        self._id = df.register_iter(self)
+        self._done = False
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """
+        Returns:
+            (`Tuple[Any, ...]`)
+                In the normal case, the iterator will return a `Tuple` at each call.
+            (`None`)
+                In the case that the `DataFrame` is not sealed and the new rows are
+                not ready yet, the iterator will return `None`. The caller should
+                determine whether to block the iteration or exit the loop.
+        Raises:
+            (`StopIteration`)
+                The iteration end iff the `DataFrame` is sealed and the last row is
+                reached.
+        """
+        if self._done:
+            raise StopIteration
+
+        df = self._df_ref()
+
+        code, row = df.get(self._offset, count = self._batch_size, iter_id = self._id)
+
+        if code == Responses.INDEX_GC:
+            raise IndexError
+
+        elif code == Responses.INDEX_OOB_UNSEALED:
+            if self._block:
```

Comment:
Why not give iterator an `Event`, if the response is index_oob_unsealed, call event.wait(), and add a `notify()` function, the `dataframe` can call iter.notify() to wake up iterators.

Reply:
It could be like this:
```python
    elif code == Responses.INDEX_OOB_UNSEALED:
        self._event.wait()
```
---
```python
  # dataframe call iterator.notify
  for iter in self._iters:
     if iter.status == 'wait':
         iter.notify()
```

Reply:
Using an event is probably a better idea. Will still need to keep track of which index is being waited on as that one event is used both for killing the process and signifying that values are available. In dataframe v1 this was a problem because we would notify after adding one value in, and if a batch iterator was waiting, it would return None. Ill update to event design.